include/jemalloc/internal/prof_inlines_a.h

```diff
@@ -3,11 +3,36 @@
 
 #include "jemalloc/internal/mutex.h"
 
+JEMALLOC_ALWAYS_INLINE void
+prof_active_assert() {
+	cassert(config_prof);
+	/*
+	 * If opt_prof is off, then prof_active must always be off, regardless
+	 * of whether prof_active_mtx is in effect or not.
+	 */
+	assert(opt_prof || !prof_active);
+}
+
+JEMALLOC_ALWAYS_INLINE bool
+prof_active_get_unlocked(void) {
+	prof_active_assert();
+	/*
+	 * Even if opt_prof is true, sampling can be temporarily disabled by
+	 * setting prof_active to false.  No locking is used when reading
+	 * prof_active in the fast path, so there are no guarantees regarding
+	 * how long it will take for all threads to notice state changes.
+	 */
+	return prof_active;
+}
+
 static inline bool
-prof_accum_add(tsdn_t *tsdn, prof_accum_t *prof_accum,
-    uint64_t accumbytes) {
+prof_accum_add(tsdn_t *tsdn, uint64_t accumbytes) {
```

Comment:
Let's call this function `prof_idump_accum`, since that's the only purpose?

Also let's change both add and cancel to accept tsd instead of tsdn?

Reply:
I'm ending up only changing the two functions' names: `prof_accum_add()` -> `prof_idump_accum()`, and `prof_accum_cancel()` -> `prof_idump_rollback()`. (The name `prof_idump_cancel()` sounds like it attempts to cancel the dump.)

I decided to give up after changing some of the `tsdn`s into `tsd`: now `prof_idump_accum()` is only called in the event handler which is neat and easy to change, and the change for `prof_idump_rollback()` only involves a few `rallocx()` related call sites. but then I found I should also change `prof_idump()` since it's also only called in the event handler now, but after all the call sites are changed, I further found that I should probably also change `prof_gdump()` since it's essentially identical to `prof_idump()` and should also only be executed when a `tsd` is available, and then I figured that a lot of other places in the profiling codebase might need changes as well. So instead of bloating this PR, I'll stick to the `tsdn` way for now and will consider making a larger scale change in the future.

I think the discrepancy is on the level of design principle: previously, `tsdn` is used whenever the underlying implementation doesn't require a `tsd`; what we want now is that `tsd` should be used whenever it doesn't make logical sense to call the function when a `tsd` is not available. This can be a discussion on the scale of the entire jemalloc codebase.

Reply:
Sounds reasonable. Let's not touch tsdn in this PR then.